### PR TITLE
contribution: Python Stub File

### PR DESCRIPTION
**Pull request by iguessitsjeff**
PR adds a stub file to the Python Binding source code in order to give a more clear description when implementing using the Python Package.

Implemented based on [Pyo3 Documentation](https://pyo3.rs/main/python-typing-hints)

### Vscode Before
Before

### Vscode After
After

### CODE REVIEW

**Comment by stefan-steadfast** [comment]
This looks great, thanks for contributing. @iguessitsjeff is it enough to simply merge and Pyo3/Maturin will take care of the rest?

**Comment by stefan-gorules** [comment]
This looks great, thanks for contributing. @iguessitsjeff is it enough to simply merge and Pyo3/Maturin will take care of the rest?

**Comment by stefan-gorules** [comment]
This looks great, thanks for contributing. @iguessitsjeff is it enough to simply merge and Pyo3/Maturin will take care of the rest?

**Comment by iguessitsjeff** [comment]
> This looks great, thanks for contributing. @iguessitsjeff is it enough to simply merge and Pyo3/Maturin will take care of the rest?

@stefan-gorules Yes, that is all taken care of by Pyo3/Maturin. One thing that I wanted to verify is the expected return values of some of the functions, for example the evaluate functions:
https://github.com/gorules/zen/blob/master/bindings/python/src/engine.rs#L101-L102

It looks like it can return different values other than a python dictionary. Can someone help to list all the expected return values and I can convert the return value into a Union type.

**Comment by stefan-gorules** [comment]
Hi @iguessitsjeff, this has been held up slightly to make room for #185. After it's merged we can take a look at making minor changes to the stub (as we'll now provide async API for evaluation).

I'll try to provide types here in a bit in TypeScript format, as I'm not fully familiar with Python type capabilities. We can then take a look at updating Python types based on that.

File: bindings/python/zen.pyi [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from typing import Optional
+
+class ZenEngine:
+    def __init__(self, maybe_options: Optional[dict] = None) -> None: ...
+    
+    def evaluate(self, key: str, ctx: dict, opts: Optional[dict] = None) -> dict: ...
```
Comment (stefan-gorules) [inline bindings/python/zen.pyi:6]: Here in evaluate we return:

```
type EvaluateResponse = {
  performance: str;
  result: dict;
  trace: dict;
}
```

File: bindings/python/zen.pyi [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    
+    def create_decision(self) -> ZenDecision: ...
+    
+    def get_decision(self, key: str) -> ZenDecision: ...
+
+class ZenDecision:
+    def evaluate(self,ctx: dict, opts: Optional[dict] = None) -> dict: ...
```
Comment (stefan-gorules) [inline bindings/python/zen.pyi:13]: Same as above (`EvaluateResponse`)

File: bindings/python/zen.pyi [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    
+    def validate(self) -> None: ...
+    
+def evaluate_expression(expression: str, ctx: Optional[dict] = None) -> dict: ...
```
Comment (stefan-gorules) [inline bindings/python/zen.pyi:17]: This returns `Any` as it can be a boolean, number etc.

File: bindings/python/zen.pyi [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+def evaluate_unary_expression(expression: str, ctx: dict) -> bool: ...
+
+def render_template(template: str, ctx: dict) -> dict: ...
```
Comment (stefan-gorules) [inline bindings/python/zen.pyi:21]: This returns `Any` as well.

**Review by stefan-gorules — CHANGES_REQUESTED** [review]
Please also see #185 which adds two new methods:

```python
ZenEngine.async_evaluate()
ZenDecision.async_evaluate()
```

Here we should return `Awaitable[EvaluateResponse]`

File: bindings/python/zen.pyi (hunk 4 above)
Comment (iguessitsjeff) [reply bindings/python/zen.pyi]: Updated!

**Comment by iguessitsjeff** [comment]
Apologies for the delay!
I made the updates based on the comments.

@stefan-gorules The layout of the struct helped, I created a [TypedDict](https://peps.python.org/pep-0589/) class as there is no structs in Python. The TypedDict allows users to access the object while also getting hints from the IDE as to what keys are present.

TypedDict

Additionally, added the Async Functions with the Awaitable return type. Validated that the type hints appear as expected when using `async.gather`.

AsyncGather

And you are able to access the same keys once accessing the index of the gathered results.

AsyncTypedDict

**Review by stefan-gorules — APPROVED** [review]

**Pull request merged**